### ISSUE TITLE
Difference() behavior not consistent between joint types

### ISSUE THREAD

**Issue by stephane-caron** [stack-of-tasks/pinocchio#1752]
As of Pinocchio 2.6.4, ``pin.difference`` doesn't behave the same for different joint types when infinity is involved:

### Free flyers:

```python
>>> import pinocchio as pin
>>> model = pin.Model()
>>> model.addJoint(0, pin.JointModelFreeFlyer(), pin.SE3.Identity(), "root_joint")
1
>>> model.upperPositionLimit
array([1.79769313e+308, 1.79769313e+308, 1.79769313e+308, 1.79769313e+308,
       1.79769313e+308, 1.79769313e+308, 1.79769313e+308])
>>> pin.difference(model, pin.neutral(model), model.upperPositionLimit)
array([nan, nan, nan, nan, nan, nan])
```

### Revolute joints:

```python
>>> import pinocchio as pin
>>> model = pin.Model()
>>> model.addJoint(0, pin.JointModelRZ(), pin.SE3.Identity(), "foo")
1
>>> model.upperPositionLimit
array([1.79769313e+308])
>>> pin.difference(model, pin.neutral(model), model.upperPositionLimit)
array([1.79769313e+308])
```

### Expected behavior:

I think the second behavior is the right one, as it is expected that "+infinity - finite = +infinity". (Also, that's the one that would trivialize computing joint limits in IK, which is my current use case :wink:)

**Comment by cmastalli** [comment]
I believe it should report `inf` for undefined limits. Moreover, the result of `diff` should be also `inf`.

**Comment by jcarpent** [comment]
If you have an unbounded joint, then you should instead use the unbounded joint type, which is done for that.
The default value for the joint limits is +inf: the results you obtained come from internal computations.
We cannot guarantee a fixed output, as it really depends on the compiler in use.

I would say that it is the role of the end-user to properly use the correct input and output of the difference operations (which act on Lie groups).

**Comment by jcarpent** [comment]
For the case of FreeFlyer, the bounds on the rotation part have no meaning in general. The NaN comes from there.

**Comment by stephane-caron** [comment]
OK, I see the stakes better. I agree it would make sense to have well-formed configurations as a pre-requisite for ``difference``. Then the next thing to do would be to document this assumption, e.g. with a ``@note`` stating that the function's behavior is unspecified otherwise. (Ideally the type system could actually check this assumption for us, rather than delegating the task to end users.)

By (ii), do you mean changing the type of position limits to be something more structured than a configuration vector?

The use case I have in mind are joint position limits in a velocity-based IK. Currently it relies on

```python
Delta_q_max = pin.difference(configuration.model, q_act, q_max)
```

And then some ad-hoc parsing on NaNs/big floats to figure out which tangent coordinates are not proper limits. One way to get rid of this would be to add something like a ``JointModel.hasLimits()`` function, so that we can easily construct the slice of tangent coordinates to look at.

**Comment by jcarpent** [comment]
Good idea to add the notion of hasConfigurationLimits() for each joint.
@fabinsch Could you handle this feature and extend the documentation accordingly?

**Comment by jcarpent** [comment]
About this notion of limits, a FreeFlyer will have a bound on the translation but not on the rotation.

**Comment by stephane-caron** [comment]
> One way to get rid of this would be to add something like a JointModel.hasLimits() function, so that we can easily construct the slice of tangent coordinates to look at.

I should have made it explicit, but what I have in mind are runtime limits, *i.e.* whether the user/URDF specified a limit explicitly.

**Comment by stephane-caron** [comment]
That works with configurations, the question is how to do the same with tangent vectors?

```python
slice = [joint.idx_v for joint in model.joints if joint.has_runtime_configuration_limits]
```

Exposing ``min_config`` and ``max_config`` would do the job :+1:

**Comment by jcarpent** [comment]
This is in the Model, no?

**Comment by jcarpent** [comment]
I don't understand your approach in fact. @stephane-caron Do you think is the right one?

**Comment by jcarpent** [comment]
What we can do for sure, is to set the real bounds on the configuration for non-Euclidian manifolds.

**Comment by jcarpent** [comment]
@stephane-caron To move forward in the current issue, I would suggest you clamp the limit values in the set [-1e20, 1e20] for instance to get correct output from `difference`.

**Comment by stephane-caron** [comment]
Yep, that's in essence how [joint limits are implemented in Pink v0.5](https://github.com/tasts-robots/pink/blob/510a702996411a4528b274b4351d53a20760a4be/pink/limits.py#L30). What's not great about this approach is that in relies on how Pinocchio is implemented, rather than on its API.

To keep things in scope, let me spin a separate issue on this point: https://github.com/stack-of-tasks/pinocchio/issues/1758

From my side, the initial point of this issue about the API for ``difference`` has been addressed https://github.com/stack-of-tasks/pinocchio/pull/1753 and we could close it. :ok_hand:

**Comment by jcarpent** [comment]
> What's not great about this approach is that in relies on how Pinocchio is implemented, rather than on its API.

I don't understand this point.

**Comment by stephane-caron** [comment]
Arh, sorry you meant clamping the _inputs_ to ``difference`` to get correct outputs, while I was looking at clamping its outputs.

Clamping the inputs is not sufficient in itself because ``difference`` is not meant to handle out-of-manifold rotation coordinates. For example:

```python
>>> model = pin.Model()
>>> model.addJoint(0, pin.JointModelFreeFlyer(), pin.SE3.Identity(), "root_joint")
1
>>> pin.difference(model, pin.neutral(model), np.array([1e20] * 7))
array([9618.35346861, 9618.35346861, 9618.35346861,    0.        ,
          0.        ,    0.        ])
```

But if we make sure unbounded rotations have proper coordinates, then we get correct outputs for the other coordinates:

```python
>>> pin.difference(model, pin.neutral(model), np.array([1e20]*3 + [0, 0, 0, 1]))
array([1.e+20, 1.e+20, 1.e+20, 0.e+00, 0.e+00, 0.e+00])
```

No worries re. my issue which found a pleasant solution in https://github.com/tasts-robots/pink/pull/12. I hope our exchange here was not too unstructured and somehow useful feedback for Pinocchio :smiley:

**Comment by wxmerkt** [comment]
@stephane-caron The first example does not have the quaternion normalised which is likely why things break in unexpected ways. It's assumed that the quaternion being passed in has been normalised before

**Comment by jcarpent** [comment]
We need to fix in Pinocchio the proper value for the bounds limit for unbounded joints.
We will do that with the neutral element of the given joint type.
